src/compiler/scala/tools/nsc/ast/TreeDSL.scala

```diff
@@ -140,7 +140,18 @@ trait TreeDSL {
     def NEW(tpt: Tree, args: Tree*): Tree   = New(tpt, List(args.toList))
 
     def NOT(tree: Tree)   = Select(tree, Boolean_not)
-    def AND(guards: Tree*) = if (guards.isEmpty) EmptyTree else guards reduceLeft gen.mkAnd
+    def AND(guards: Tree*) = {
+      def binaryTreeAnd(tests: Seq[Tree]): Tree = tests match{
+        case List() => ???
+        case List(single) => single
+        case multiple =>
+          val (before, after) = multiple.splitAt(tests.size / 2)
+          gen.mkAnd(binaryTreeAnd(before), binaryTreeAnd(after))
+      }
+
+      assert(guards.nonEmpty, "Need at least one tree to AND() together, not zero")
```

Comment:
This case never happens AFAICT, so better to fail loudly if it does rather than returning a rubbish AST node

Reply:
"AFAICT"?  The types allow it and there's nothing rubbish about EmptyTree.  Could you add that back?  I'd rather that than an `assert` and a `???`.

Reply:
AFAICT = As Far As I Can Tell. I grepped for `AND()` in the codebase and there is only one callsite (this one), and it doesn't allow an empty Seq to be passed in.

The `EmptyTree` branch is basically dead code that is never exercised, so who knows if it does what someone wants if it gets inadvertently hit? Wouldn't it be better to fail loudly rather than proceeding with an untested unexercised branch that nobody has ever thought about? If we really want to put something there, why not put a literal `true`, since that's the monoid zero of the `&&` operation?

That's my reasoning anyway. If you still think we should put the `EmptyTree` back in, I'm happy to do so.

Reply:
Arbitrary plugins can exercise arbitrary code, as I recently learned again to my chagrin.

This fix is similar to the flatter ast for huge interpolations.

I wonder if there is a possible interaction with constant folding and optimization. `if (DEBUG && ...)`

Reply:
> The `EmptyTree` branch is basically dead code that is never exercised, so who knows if it does what someone wants if it gets inadvertently hit? Wouldn't it be better to fail loudly rather than proceeding with an untested unexercised branch that nobody has ever thought about?

It's a toss-up between returning something one could call reasonable vs loudly and clearly failing early on unexpected input.  Given this is out of scope for the objective here, I'd opt for the status quo (EmptyTree).

Even better I'm not against making AND take `(guard: Tree, guards: Tree*)` and avoid this dilemma entirely.

Reply:
Actually don't mind this returning `TRUE` (above in the file) on the empty case, instead.